include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -574,23 +575,23 @@ struct __get_sycl_range
         const auto __n = ::std::min(decltype(__size)(__last - __first), __size);
         assert(__offset + __n <= __size);
 
-        return __range_holder<oneapi::dpl::__ranges::all_view<value_type, AccMode>>{
-            oneapi::dpl::__ranges::all_view<value_type, AccMode>(__first.get_buffer() /* buffer */,
-                                                                 __offset /* offset*/, __n /* size*/)};
+        return __range_holder<oneapi::dpl::__ranges::all_view<value_type, _LocalAccMode>>{
+            oneapi::dpl::__ranges::all_view<value_type, _LocalAccMode>(__first.get_buffer() /* buffer */,
+                                                                       __offset /* offset*/, __n /* size*/)};
     }
 
     //SFINAE-overload for a contiguous host iterator
-    template <typename _Iter>
+    template <sycl::access::mode _LocalAccMode, typename _Iter>
     auto
-    operator()(_Iter __first, _Iter __last)
+    __process_input_iter(_Iter __first, _Iter __last)
         -> ::std::enable_if_t<is_temp_buff<_Iter>::value && __is_addressable_v<_Iter> && !is_zip<_Iter>::value &&
                                   !is_permutation<_Iter>::value,
-                              __range_holder<oneapi::dpl::__ranges::all_view<val_t<_Iter>, AccMode>>>
+                              __range_holder<oneapi::dpl::__ranges::all_view<val_t<_Iter>, _LocalAccMode>>>
     {
         using _T = val_t<_Iter>;
 
-        return __process_host_iter_impl(__first, __last, [&]() {
-            if constexpr (__is_copy_direct)
+        return __process_host_iter_impl<_LocalAccMode>(__first, __last, [&]() {
+            if constexpr (__is_copy_direct_v<_LocalAccMode>)
             {
                 //wait and copy on a buffer destructor; an exclusive access buffer, good performance
```

Comment:
I propose to add before this line assertion check:
`assert(!__is_copy_back_v<_LocalAccMode>);`

Reply:
I don't understand why this is needed.  Can you explain further?
I think  `__is_copy_back_v<_LocalAccMode> == true` is a valid state here, (and it will be handled below with `__buf.set_write_back(__is_copy_back_v<_LocalAccMode>);` within `__process_host_iter_impl()`

Reply:
In the next function we have
```C++
                if constexpr (__is_copy_back_v<_LocalAccMode>)
                    __buf.set_final_data(__first); //SYCL API for non-contiguous iterators
```
In this place this code isn't required, because by semantic `__is_copy_back_v<_LocalAccMode>` is `false`.
I propose to add assertion for this case.

Reply:
I'm not seeing how `__is_copy_back_v<_LocalAccMode>` is `false` by semantic here.  It could be either by my understanding.  Copy back is dependent on the access mode, and has nothing to do with if the input data is contiguous, so we can't determine that copy back isn't required based on this.  We don't special case based on this here because it isn't required with this `sycl::buffer` constructor.

This function is for contiguous host iterators, so it is able to utilize the host pointer + range constructor for `sycl::buffer`, rather than the host container `sycl::buffer` (which requires an extra host side allocation and copy of the data).

For this `sycl::buffer`, constructor with the host pointer, the "copy back" can merely be provided by the `__buf.set_write_back(__is_copy_back_v<_LocalAccMode>);` which is called when this lambda returns in `__process_host_iter_impl()`.  For the other non-contiguous version, it requires the extra step of `__buf.set_final_data(__first)`, because it is using the host container `sycl::buffer` constructor.  This tells the `sycl::buffer` that when it is destroyed, it should copy the data back to a specific host iterator.  

For the host pointer constructor, it just uses the supplied memory address as its host side memory, so no extra host side copy back actually occurs (after the device -> host copy).

@MikeDvorskiy  does this agree with your understanding as well?

Reply:
Yes, `__is_copy_back_v<_LocalAccMode> == true `is a valid state here.

The fact that we don't call set_final_data doesn't  tell that copying will no be, because a buffer with exclusive access does it by default (according to the spec).
Probably, it makes sense to add a detailed comment why buffer with exclusive access has different behavior and  does copy back on destructor by default while buffer(it1, it2) doesn't. Dan, could you please to add such additional comment?

Reply:
Hope @MikeDvorskiy will propose some comment about this for insert into the code.

Reply:
Sure I can add these comments.